common/utils/include/claragenomics/utils/buffer.hpp

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * NVIDIA CORPORATION and its licensors retain all intellectual property
+ * and proprietary rights in and to this software, related documentation
+ * and any modifications thereto.  Any use, reproduction, disclosure or
+ * distribution of this software and related documentation without an express
+ * license agreement from NVIDIA CORPORATION is strictly prohibited.
+ */
+
+#pragma once
+#include <claragenomics/utils/cudautils.hpp>
+#include <claragenomics/utils/allocator.hpp>
+
+namespace claragenomics
+{
+
+template <typename T, typename Allocator>
+class buffer
+{
+public:
+    using value_type     = T;
+    using size_type      = std::size_t;
+    using iterator       = value_type*;
+    using const_iterator = const value_type*;
+
+    buffer(const buffer& other) = delete;
+
+    buffer& operator=(const buffer& other) = delete;
+
+    buffer(size_type n                          = 0,
+           std::shared_ptr<Allocator> allocator = std::make_shared<Allocator>(),
+           cudaStream_t stream                  = 0)
+        : _size(n)
+        , _capacity(n)
+        , _data(nullptr)
+        , _stream(stream)
+        , _allocator(allocator)
+    {
```

Comment:
Currently, `Anchor` and `Overlap` are not trivial types due to the presence of custom constructors and non-static members with default initializers, though they can be made trivial with some effort in the future release.

Reply:
The problem is that this container does not support non-trivial types, because it will not call the destructor of the elements on destruction or shrinking.

Reply:
`Overlap` should be trivial, right? I don't see why `Anchor` requires the empty default constructor. I don't remember why we added it. I'll ask.

Reply:
Changed `Overlap` and `Anchor` to trivial types and added the check to ensure that buffer only supports trivial types and classes.